cmd/notation/internal/plugin/plugin.go

```diff
@@ -0,0 +1,81 @@
+// Copyright The Notary Project Authors.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plugin
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/notaryproject/notation/internal/httputil"
+)
+
+// MaxPluginSourceBytes specifies the limit on how many bytes are allowed in the
+// server's response to the download from URL request.
+//
+// The plugin source size must be strictly less than this value.
+var MaxPluginSourceBytes int64 = 256 * 1024 * 1024 // 256 MiB
+
+// PluginSourceType is an enum for plugin source
+type PluginSourceType int
+
+const (
+	// PluginSourceTypeFile means plugin source is file
+	PluginSourceTypeFile PluginSourceType = 1 + iota
+
+	// PluginSourceTypeURL means plugin source is URL
+	PluginSourceTypeURL
+)
+
+const (
+	// MediaTypeZip means plugin file is zip
+	MediaTypeZip = "application/zip"
+
+	// MediaTypeGzip means plugin file is gzip
+	MediaTypeGzip = "application/x-gzip"
+)
+
+// DownloadPluginFromURL downloads plugin file from url to a tmp directory
+func DownloadPluginFromURL(ctx context.Context, pluginURL string, tmpFile io.Writer) error {
+	// Get the data
+	client := httputil.NewAuthClient(ctx, &http.Client{Timeout: 100 * time.Second})
```

Comment:
Also if we are doing 100 seconds timeout; we need to communicate this to user otherwise user might think the command is stuck

Reply:
@shizhMSFT The purpose of having a timeout is to avoid scripts and CI/CD pipelines hang at this step indefinitely. Golang provides a timeout with 0 as default value, I found the `100 seconds` from here: https://learn.microsoft.com/en-us/dotnet/api/system.net.http.httpclient.timeout?view=net-8.0#remarks.

Reply:
> Also if we are doing 100 seconds timeout; we need to communicate this to user otherwise user might think the command is stuck

@priteshbandi The default value is 0 which means no timeout. Adding this value actually prevent user from stuck.

Reply:
Can you make it 20 minutes and document it somewhere?

Reply:
Updated to 20 minutes, added the info in the command description.

Reply:
For all practical purpose waiting more than 2 mins without showing progress is basically stuck.

Reply:
> For all practical purpose waiting more than 2 mins without showing progress means that process is stuck.

@priteshbandi I had a discussion with @shizhMSFT, updated the timeout to 10 minutes. Stated the value in the description of the `--url` flag. If user has a very slow network connection, their workaround is to download the file then install from file system.

Reply:
My concern still persist. Can we tell customers that we are downloading plugin, probably using progress bar(if possible) ?

Reply:
> My concern still exists. Can we tell customers that we are downloading plugin, probably using progress bar(if possible) ?

@priteshbandi Adding progress bar is considered as a new feature which is not in this PR's scope. I added printouts before and after the download so that user could know what is happening.